src/jvm/main/org/jetbrains/kotlinx/lincheck/execution/ExecutionScenario.kt

```diff
@@ -0,0 +1,74 @@
+/*
+ * #%L
+ * Lincheck
+ * %%
+ * Copyright (C) 2015 - 2018 Devexperts, LLC
+ * %%
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Lesser Public License for more details.
+ *
+ * You should have received a copy of the GNU General Lesser Public
+ * License along with this program.  If not, see
+ * <http://www.gnu.org/licenses/lgpl-3.0.html>.
+ * #L%
+ */
+package org.jetbrains.kotlinx.lincheck.execution
+
+import org.jetbrains.kotlinx.lincheck.Actor
+import org.jetbrains.kotlinx.lincheck.appendExecutionScenario
+import java.util.stream.Stream
+
+/**
+ * This class represents an execution scenario, which
+ * is generated by an [ExecutionGenerator] and then \
+ * used by a [Strategy] which produces an [ExecutionResult].
+ */
+class ExecutionScenario(
+        /**
+         * The initial sequential part of the execution.
+         * It helps to produce different initial states
+         * before the parallel part.
+         *
+         * The initial execution part should contain only non-suspendable actors;
+         * otherwise, the single initial execution thread will suspend with no chance to be resumed.
+         */
+        val initExecution: MutableList<Actor>,
+        /**
+         * The parallel part of the execution, which is used
+         * to find an interleaving with incorrect behaviour.
+         */
+        val parallelExecution: MutableList<MutableList<Actor>>,
+        /**
+         * The last sequential part is used to test that
+         * the data structure is in some correct state.
+         *
+         * If this execution scenario contains suspendable actors, the post part should be empty;
+         * if not, an actor could resume a previously suspended one from the parallel execution part.
+         */
+        val postExecution: MutableList<Actor>) {
+    /**
+     * Returns the number of threads used in the parallel part of this execution.
+     */
+    val threads: Int
+        get() = parallelExecution.size
+
+    /**
+     * Returns `true` if there is at least one suspendable actor in the generated scenario
+     */
+    fun hasSuspendableActors(): Boolean {
+        return Stream.concat(parallelExecution.stream().flatMap { obj: List<Actor> -> obj.stream() }, postExecution.stream()).anyMatch(Actor::isSuspendable)
```

Comment:
Please don't use Java's streams in Kotlin + you can write `fun hasSuspendableActors() = ...`